yt_dlp/networking/common.py

```diff
@@ -0,0 +1,459 @@
+    Any other exception raised will be treated as a handler issue.
+
+    If a Request is not supported by the handler, an UnsupportedRequest
+    should be raised with a reason.
+
+    By default, some checks are done on the request in _validate() based on the following class variables:
+    - `_SUPPORTED_URL_SCHEMES`: may contain a tuple of supported url schemes.
+        Any Request with an url scheme not in this list will raise an UnsupportedRequest.
+
+    - `_SUPPORTED_PROXY_SCHEMES`: may contain a tuple of support proxy url schemes. Any Request that contains
+        a proxy url with an url scheme not in this list will raise an UnsupportedRequest.
+
+    - `_SUPPORTED_FEATURES`: may contain a tuple of supported features, as defined in Features enum.
+
+    Parameters:
+    @param logger: logger instance
+    @param headers: HTTP Headers to include when sending requests.
+    @param cookiejar: Cookiejar to use for requests.
+    @param timeout: Socket timeout to use when sending requests.
+    @param proxies: Proxies to use for sending requests.
+    @param source_address: Client-side IP address to bind to for requests.
+    @param verbose: Print debug traffic
+    @param prefer_system_certs: Whether to prefer system certificates over other means (e.g. certifi).
+    @param client_cert: SSL client certificate configuration.
+            Tuple of (client_certificate, client_certificate_key, client_certificate_password).
+    @param verify: Verify SSL certificates
+    @param legacy_ssl_support: Enable legacy SSL options such as legacy server connect and older cipher support.
+
+    Some configuration options may be available for individual Requests too. In this case,
+    either the Request configuration option takes precedence or they are merged.
+
+    Requests may have additional optional parameters defined as extensions.
+     RequestHandler subclasses may choose to support custom extensions.
+
+    The following extensions are defined for RequestHandler:
+    - `cookiejar`: Cookiejar to use for this request
+    - `timeout`: socket timeout to use for this request
+
+    Apart from the url protocol, proxies dict may contain the following keys:
+    - `all`: proxy to use for all protocols. Used as a fallback if no proxy is set for a specific protocol.
+    - `no`: comma seperated list of hostnames (optionally with port) to not use a proxy for.
+    Note: a RequestHandler may not support these, as defined in `_SUPPORTED_FEATURES`.
+
+    """
+
+    _SUPPORTED_URL_SCHEMES: tuple = None
+    _SUPPORTED_PROXY_SCHEMES: tuple = None
+    _SUPPORTED_FEATURES: tuple = None
+
+    def __init__(
+        self,
+        *,
+        logger=None,  # TODO
+        headers: CaseInsensitiveDict = None,
+        cookiejar: CookieJar = None,
+        timeout: float | int | None = None,
+        proxies: dict = None,
+        source_address: str = None,
+        verbose: bool = False,
```

Comment:
We need to know if we have debug traffic on. It may not all be able to go through the logger e.g http lib connection traffic for urllib and requests which just prints to stdout

Reply:
Then I would make it the responsibility of the director/handler/... to log and not introduce a new method on `Logger`? We can create a helper function in networking utils for logging debug traffic, like
```py
logger.debug(format_debug_traffic(data))
```

Reply:
possibly

The cases I've seen are:

- manually write a message if debugtraffic is set (in request director or handler)
- enabled debug for HTTPConnection(?) if debugtraffic is set, which prints to stdout (urllib and urllib3)
- print debug logging from urllib3 if debutraffic is set (urllib3 uses the `logging` module)